File: Heeseop An/2023-02-07-Linked-List-Cycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        seen = set()

        while head:
            if head in seen:
                return True
            else:
                seen.add(head)

            head = head.next

        return False

File: Dave Min/2023-02-07-LinkedListCycle.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution
{
    public boolean hasCycle(ListNode head)
    {
        if (head == null)
            return false;
        ListNode turtle = head;
        ListNode rabbit = head.next;

        while (turtle != rabbit)
        {
            if (rabbit == null || rabbit.next == null)
                return false;

            turtle = turtle.next;
            rabbit = rabbit.next.next;
        }
        return true;
    }
}